src/xb-stack-private.h

```diff
@@ -8,10 +8,50 @@
 
 #include <glib-object.h>
 
+#include "xb-opcode-private.h"
 #include "xb-stack.h"
 
 G_BEGIN_DECLS
 
+/* Members of this struct should not be accessed directly — use the accessor
+ * functions below instead. */
+struct _XbStack {
+	/*< private >*/
+	gint		 ref;
+	gboolean	 stack_allocated;	/* whether this XbStack was allocated with alloca() */
+	guint		 pos;	/* index of the next unused entry in .opcodes */
+	guint		 max_size;
+	XbOpcode	 opcodes[];	/* allocated as part of XbStack */
+};
+
+/**
+ * xb_stack_new_inline:
+ * @max_stack_size: maximum size of the stack
+ *
+ * Creates a stack for the XbMachine request. Only #XbOpcodes can be pushed and
+ * popped from the stack.
+ *
+ * The stack will be allocated on the current C stack frame, so @max_stack_size
+ * should be chosen to not overflow the C process’ stack.
+ *
+ * Returns: (transfer full): a #XbStack
+ *
+ * Since: 0.3.1
+ **/
+#define xb_stack_new_inline(max_stack_size) \
+(G_GNUC_EXTENSION ({ \
+	/* This function has to be static inline so we can use g_alloca(), which \
```

Comment:
I'm a bit unclear on why it needs to be static inline. Got any background to make me sound like a Unix newbie?

Reply:
oops, that comment is out of date — it’s a macro, rather than `static inline`. Making it `static inline` doesn’t work.

It needs to be a macro because allocations made with `alloca()` live as long as the stack frame they’re allocated in, so the only way to abstract them is to move them into a macro.

Reply:
I was going to say "why can't we just do XbStack stack = XB_STACK_INIT" but then that's basically what this is. Perhaps we could make it more like XB_QUERY_CONTEXT_INIT in use rather than pretending it's an action function.

Reply:
I’m not sure that would make things clearer, since `XB_QUERY_CONTEXT_INIT` follows the pattern for stack-allocated, non-pointer variables. `alloca()` is a weird one in that it creates a stack-allocated, pointer variable.

Reply:
Ya, indeed. I'm not sure why I'm getting so uptight, this is just internal non-exported API.

Reply:
It’s good to get things right even for internal API, thanks for checking! :slightly_smiling_face: